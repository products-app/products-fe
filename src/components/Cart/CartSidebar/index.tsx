import Sidebar from '@/components/common/Sidebar'
import NotFoundCartItems from '@/components/NotFound/CartItems'
import ListCart from '@/components/Cart/CartList'
import { useCart } from '../hook'
import {
  SidebarContainer,
  CartContainer,
  ItemsContainer,
  ItemsText,
  CartControl,
  ButtonContainer,
} from './styles'

const CartSidebar = () => {
  const {
    cartItems,
    totalCartItems,
    increaseItem,
    decreaseItem,
    removeItem,
    openSidebar,
    handleCloseSidebar,
    handleOpenCheckout,
  } = useCart()

  return (
    <Sidebar isOpen={openSidebar} onClose={handleCloseSidebar}>
      <SidebarContainer>
        <CartContainer>
          <ItemsText size="lg">Carrinho de compras</ItemsText>
          <ItemsContainer id="cart-sidebar-items">
            {cartItems && totalCartItems > 0 && (
              <ListCart
                items={cartItems}
                actions={{ decreaseItem, increaseItem, removeItem }}
              />
            )}

            {totalCartItems === 0 && <NotFoundCartItems />}
          </ItemsContainer>
        </CartContainer>

        {totalCartItems > 0 && (
          <CartControl>
            <ButtonContainer variant="primary" onClick={handleOpenCheckout}>
              Comprar agora
            </ButtonContainer>
          </CartControl>
        )}
      </SidebarContainer>
    </Sidebar>
  )
}

export default CartSidebar
