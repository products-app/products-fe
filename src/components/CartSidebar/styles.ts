import { styled, Button, Text, Card } from '@lebernardo/react'

export const SidebarContainer = styled('nav', {
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  height: '100%',
  position: 'relative',
  padding: '24px',
})

export const CartContainer = styled('div', {
  height: '93%',
  position: 'relative',
  width: '100%',
})

export const ItemsText = styled(Text, {
  marginBottom: '1.25rem',
})

export const ItemsContainer = styled('div', {
  width: '100%',
  height: '90%',
  position: 'relative',
  overflowY: 'scroll'
})

export const CartControl = styled(Card, {
  position: 'fixed',
  right: 0,
  bottom: 0,
  width: '100%',
})

export const ButtonContainer = styled(Button, {
  margin: '1rem 0',
  width: '100%',
})
