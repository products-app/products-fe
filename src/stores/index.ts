import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'
import { createUserSlice } from './slices/userSlice'
import { createCartSlice } from './slices/cartSlice'
import { createAppSlice } from './slices/appSlice'

export const useStore = create(
  persist<app.ICart & app.IUser & app.IAppControl>(
    (set, get) => ({
      ...createUserSlice(set),
      ...createCartSlice(set, get),
      ...createAppSlice(set),
    }),
    {
      name: 'store',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
