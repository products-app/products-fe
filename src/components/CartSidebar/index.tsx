import { countCartItems } from '@/helpers/cart'
import Sidebar from '@/components/Sidebar'
import NotFoundCartItems from '@/components/NotFound/CartItems'
import ListCart from './components/ListCart'
import { useRouter } from 'next/navigation'
import { useStore } from '@/stores'
import {
  SidebarContainer,
  CartContainer,
  ItemsContainer,
  ItemsText,
  CartControl,
  ButtonContainer,
} from './styles'

const CartSidebar = () => {
  const router = useRouter()
  const { cartItems, openSidebar, setOpenSidebar, setOpenCheckout, userToken } =
    useStore((state) => ({
      cartItems: state.items,
      openSidebar: state.openSidebar,
      setOpenSidebar: state.setOpenSidebar,
      setOpenCheckout: state.setOpenCheckout,
      userToken: state.user?.token,
    }))
  const totalCartItems = countCartItems(cartItems || [])

  const handleCloseSidebar = () => {
    setOpenSidebar(false)
  }

  const handleOpenCheckout = () => {
    if (userToken) {
      setOpenCheckout(true)
    } else {
      router.push('/account')
    }
  }

  return (
    <Sidebar isOpen={openSidebar} onClose={handleCloseSidebar}>
      <SidebarContainer>
        <CartContainer>
          <ItemsText size="lg">Carrinho de compras</ItemsText>
          <ItemsContainer id="cart-sidebar-items">
            {cartItems && totalCartItems > 0 && <ListCart items={cartItems} />}

            {totalCartItems === 0 && <NotFoundCartItems />}
          </ItemsContainer>
        </CartContainer>

        {totalCartItems > 0 && (
          <CartControl>
            <ButtonContainer variant="primary" onClick={handleOpenCheckout}>
              Comprar agora
            </ButtonContainer>
          </CartControl>
        )}
      </SidebarContainer>
    </Sidebar>
  )
}

export default CartSidebar
