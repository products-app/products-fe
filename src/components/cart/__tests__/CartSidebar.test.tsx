import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { useStore } from '@/stores'
import CartSidebar from '../CartSidebar/index'

const cartItems = [
  {
    id: 1,
    name: 'Product 1',
    price: 29.9,
    stock: 10,
    quantity: 3,
  },
]
const initialState = useStore.getState()

describe('Cart Sidebar component', () => {
  afterEach(() => {
    jest.restoreAllMocks()
  })

  test('should render correctly', () => {
    useStore.setState({ ...initialState, items: cartItems }, true)

    const { container } = render(<CartSidebar />)

    expect(screen.getByText('Product 1')).toBeInTheDocument()
    expect(() => screen.getByText('no exist')).toThrow()
    expect(container).toMatchSnapshot()
  })

  test('should render a empty cart correctly', () => {
    useStore.setState({ ...initialState, items: [] }, true)

    const { container } = render(<CartSidebar />)

    expect(screen.queryByText('Product 1')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('should render correctly when cart items is undefined', () => {
    useStore.setState({ ...initialState, items: undefined }, true)

    const { container } = render(<CartSidebar />)

    expect(screen.queryByText('Product 1')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  test('should redirect to the login page when click on button "Buy now"', async () => {
    useStore.setState({ ...initialState, items: cartItems }, true)
    render(<CartSidebar />)

    const handleClick = jest.fn()
    const buyButtonElement = screen.getByRole('button', {
      name: /Comprar agora/i,
    })
    buyButtonElement.addEventListener('click', handleClick, false)

    await userEvent.click(buyButtonElement)

    expect(handleClick).toHaveBeenCalled()
  })

  test('should set open sidebar when click on button "Buy now"', async () => {
    useStore.setState(
      {
        ...initialState,
        items: cartItems,
        user: { token: '123', id: 1, name: 'Maria' },
      },
      true,
    )
    render(<CartSidebar />)

    const handleClick = jest.fn()
    const buyButtonElement = screen.getByRole('button', {
      name: /Comprar agora/i,
    })
    buyButtonElement.addEventListener('click', handleClick, false)

    await userEvent.click(buyButtonElement)

    expect(handleClick).toHaveBeenCalled()
  })

  test('should close sidebar', async () => {
    useStore.setState({ ...initialState, items: cartItems }, true)
    render(<CartSidebar />)

    const handleClick = jest.fn()
    const buyButtonElement = screen.getByRole('button', {
      name: 'button-close-sidebar',
    })

    buyButtonElement.addEventListener('click', handleClick, false)

    await userEvent.click(buyButtonElement)

    expect(handleClick).toHaveBeenCalled()
  })
})
