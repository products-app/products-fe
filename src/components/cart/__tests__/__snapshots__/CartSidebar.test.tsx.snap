// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Sidebar component should render a empty cart correctly 1`] = `
<div>
  <section
    class="c-hEQxdV"
    id="sidebar"
    style="transform: translateX(100%);"
  >
    <button
      aria-label="button-close-sidebar"
      class="c-eOLtzh"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="56"
          y2="200"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="200"
          y2="56"
        />
      </svg>
    </button>
    <nav
      class="c-nMCuV"
    >
      <div
        class="c-gzevCw"
      >
        <p
          class="c-dgMvMW c-fMsmQc c-dgMvMW-eQRsLC-size-lg"
        >
          Carrinho de compras
        </p>
        <div
          class="c-jwMKId"
          id="cart-sidebar-items"
        >
          <div
            class="c-dzUtkR"
          >
            <img
              alt="Imagem de carrinho vazio"
              data-nimg="1"
              decoding="async"
              height="0"
              loading="lazy"
              sizes="100vw"
              src="/_next/image?url=%2Fempty-cart.png&w=3840&q=75"
              srcset="/_next/image?url=%2Fempty-cart.png&w=640&q=75 640w, /_next/image?url=%2Fempty-cart.png&w=750&q=75 750w, /_next/image?url=%2Fempty-cart.png&w=828&q=75 828w, /_next/image?url=%2Fempty-cart.png&w=1080&q=75 1080w, /_next/image?url=%2Fempty-cart.png&w=1200&q=75 1200w, /_next/image?url=%2Fempty-cart.png&w=1920&q=75 1920w, /_next/image?url=%2Fempty-cart.png&w=2048&q=75 2048w, /_next/image?url=%2Fempty-cart.png&w=3840&q=75 3840w"
              style="color: transparent;"
              width="0"
            />
            <p
              class="c-dgMvMW c-igsGei c-dgMvMW-kWyXas-size-md"
            >
              Adicione algo em seu carrinho...
            </p>
          </div>
        </div>
      </div>
    </nav>
  </section>
</div>
`;

exports[`Cart Sidebar component should render correctly 1`] = `
<div>
  <section
    class="c-hEQxdV"
    id="sidebar"
    style="transform: translateX(100%);"
  >
    <button
      aria-label="button-close-sidebar"
      class="c-eOLtzh"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="56"
          y2="200"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="200"
          y2="56"
        />
      </svg>
    </button>
    <nav
      class="c-nMCuV"
    >
      <div
        class="c-gzevCw"
      >
        <p
          class="c-dgMvMW c-fMsmQc c-dgMvMW-eQRsLC-size-lg"
        >
          Carrinho de compras
        </p>
        <div
          class="c-jwMKId"
          id="cart-sidebar-items"
        >
          <ul>
            <li
              class="c-kMCjoS"
            >
              <div
                class="c-dZtilx"
              >
                <img
                  alt="Product 1"
                  data-nimg="1"
                  decoding="async"
                  height="0"
                  loading="lazy"
                  sizes="100vw"
                  src="/_next/image?url=%2Fproduct-not-found.jpg&w=3840&q=75"
                  srcset="/_next/image?url=%2Fproduct-not-found.jpg&w=640&q=75 640w, /_next/image?url=%2Fproduct-not-found.jpg&w=750&q=75 750w, /_next/image?url=%2Fproduct-not-found.jpg&w=828&q=75 828w, /_next/image?url=%2Fproduct-not-found.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fproduct-not-found.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fproduct-not-found.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fproduct-not-found.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fproduct-not-found.jpg&w=3840&q=75 3840w"
                  style="color: transparent;"
                  width="0"
                />
              </div>
              <figure
                class="c-hBSDYn"
              >
                <p
                  class="c-dgMvMW c-dgMvMW-fAKXXr-size-sm"
                >
                  Product 1
                </p>
                <p
                  class="c-dgMvMW c-fFOkxT c-dgMvMW-cEPjnm-size-xs"
                >
                  Qtde:
                   
                  <strong
                    aria-label="item-quantity"
                    data-cy="item-quantity"
                  >
                    3
                  </strong>
                </p>
              </figure>
              <div
                class="c-gsXdEb"
              >
                <p
                  class="c-dgMvMW c-dnNpSg c-dgMvMW-kWyXas-size-md"
                >
                  R$Â 89,70
                </p>
                <div
                  class="c-gJlbUq"
                >
                  <button
                    class="c-gXmMMa"
                    data-cy="decrease-item"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      viewBox="0 0 256 256"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <rect
                        fill="none"
                        height="256"
                        width="256"
                      />
                      <line
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="16"
                        x1="40"
                        x2="216"
                        y1="128"
                        y2="128"
                      />
                    </svg>
                  </button>
                  <button
                    class="c-fOZJzt"
                    data-cy="increase-item"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      viewBox="0 0 256 256"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <rect
                        fill="none"
                        height="256"
                        width="256"
                      />
                      <line
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="16"
                        x1="40"
                        x2="216"
                        y1="128"
                        y2="128"
                      />
                      <line
                        fill="none"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="16"
                        x1="128"
                        x2="128"
                        y1="40"
                        y2="216"
                      />
                    </svg>
                  </button>
                </div>
              </div>
              <div
                class="c-gsXdEb"
              >
                <button
                  class="c-cHEnSC"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    viewBox="0 0 256 256"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <rect
                      fill="none"
                      height="256"
                      width="256"
                    />
                    <line
                      fill="none"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="16"
                      x1="216"
                      x2="40"
                      y1="56"
                      y2="56"
                    />
                    <line
                      fill="none"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="16"
                      x1="88"
                      x2="168"
                      y1="24"
                      y2="24"
                    />
                    <path
                      d="M200,56V208a8,8,0,0,1-8,8H64a8,8,0,0,1-8-8V56"
                      fill="none"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="16"
                    />
                  </svg>
                </button>
              </div>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="c-bfdXkU c-faTUNr"
      >
        <button
          class="c-gJqLDS c-lnvOQW c-gJqLDS-dsZVoc-variant-primary c-gJqLDS-bIqJQp-size-medium"
        >
          Comprar agora
        </button>
      </div>
    </nav>
  </section>
</div>
`;

exports[`Cart Sidebar component should render correctly when cart items is undefined 1`] = `
<div>
  <section
    class="c-hEQxdV"
    id="sidebar"
    style="transform: translateX(100%);"
  >
    <button
      aria-label="button-close-sidebar"
      class="c-eOLtzh"
    >
      <svg
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="56"
          y2="200"
        />
        <line
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="200"
          x2="56"
          y1="200"
          y2="56"
        />
      </svg>
    </button>
    <nav
      class="c-nMCuV"
    >
      <div
        class="c-gzevCw"
      >
        <p
          class="c-dgMvMW c-fMsmQc c-dgMvMW-eQRsLC-size-lg"
        >
          Carrinho de compras
        </p>
        <div
          class="c-jwMKId"
          id="cart-sidebar-items"
        >
          <div
            class="c-dzUtkR"
          >
            <img
              alt="Imagem de carrinho vazio"
              data-nimg="1"
              decoding="async"
              height="0"
              loading="lazy"
              sizes="100vw"
              src="/_next/image?url=%2Fempty-cart.png&w=3840&q=75"
              srcset="/_next/image?url=%2Fempty-cart.png&w=640&q=75 640w, /_next/image?url=%2Fempty-cart.png&w=750&q=75 750w, /_next/image?url=%2Fempty-cart.png&w=828&q=75 828w, /_next/image?url=%2Fempty-cart.png&w=1080&q=75 1080w, /_next/image?url=%2Fempty-cart.png&w=1200&q=75 1200w, /_next/image?url=%2Fempty-cart.png&w=1920&q=75 1920w, /_next/image?url=%2Fempty-cart.png&w=2048&q=75 2048w, /_next/image?url=%2Fempty-cart.png&w=3840&q=75 3840w"
              style="color: transparent;"
              width="0"
            />
            <p
              class="c-dgMvMW c-igsGei c-dgMvMW-kWyXas-size-md"
            >
              Adicione algo em seu carrinho...
            </p>
          </div>
        </div>
      </div>
    </nav>
  </section>
</div>
`;
